server/src/main/java/org/elasticsearch/rest/RestController.java

```diff
@@ -218,7 +219,8 @@ public void dispatchBadRequest(final RestChannel channel, final ThreadContext th
         }
     }
 
-    private void dispatchRequest(RestRequest request, RestChannel channel, RestHandler handler) throws Exception {
+    private void dispatchRequest(RestRequest request, RestChannel channel, RestHandler handler,
+                                 ThreadContext threadContext) throws Exception {
```

Comment:
technically we don't need to modify this method's signature since the class has the NodeClient, so we could do `client.threadPool().getThreadContext()` to get the ThreadContext and even store the reference in the class constructor.